src/main/java/org/jenkinsci/lib/configprovider/model/ConfigFileManager.java

```diff
@@ -115,10 +115,14 @@ public static FilePath provisionConfigFile(ConfigFile configFile, @Nullable EnvV
 
         if (configFile.isReplaceTokens()) {
             try {
-                fileContent = TokenMacro.expandAll(build, workspace, listener, fileContent);
+                // JENKINS-57417: 'env' must be processed first, as:
+                // * EnvVar.expand() / Util.replaceMacro() only handles simple macros, ${FOO}, however,
+                //   TokenMacro.expandAll() throws MacroEvaluationException's when they are present.
```

Comment:
sorry, I don't quite understand what you try to explain here; what do you mean by "...when they are present"? what must not be present?

Reply:
My exposure to the underlying TokenMacro is limited, from what I was able to trace, the [expansion model](https://wiki.jenkins.io/display/JENKINS/Token+Macro+Plugin#TokenMacroPlugin-TokenExpansionModel) supports both parameterized and unparameterized tokens; wherein the token id corresponds to a specific provider class, in this case `ENV` -> [EnvironmentVariableMacro](https://github.com/jenkinsci/token-macro-plugin/blob/master/src/main/java/org/jenkinsci/plugins/tokenmacro/impl/EnvironmentVariableMacro.java#L23). This is enforced in calls to the TokenMacro.expandAll() method, which attempts to resolve all forms `${FOO}` / `${FOO,param=...}` to a provider, throwing exceptions on failure. In contrast, the Util.replaceMacro() method is not token aware, and performs a simple, map-backed, replacements in the form `$KEY` / `${KEY}`.

The result of these two conditions is that configuration files substituting values using the form `${x}` are ambiguous between the two forms (although they are most likely intended for the replaceMacro() method).

For example:
```
x=${ENV, var="envX"} <-- is expected to be replaced by expandAll()
y=${envY} <-- is expected to be replaced by replaceMacro()
```
If `expandAll()` is invoked first, `envY` will failed to be resolved to a provider, and an exception will be raised; neither substitution will be performed.

It would be ideal if the replaceMacro() call could be removed entirely, and the extra variables could be appended to the `Run` (build), and `envY` could be refactored to be in the same form as `envX`.

Reply:
I've changed the wording of the comment to be more clear.